### PR TITLE
Tutorials for getting started with Scala.js, Vite, Laminar and ST

### PR DESCRIPTION
**Pull request by sjrd**
These tutorials are the written equivalent of recent talks I made at ScalaCon and Scala.IO, as well as videos I recorded at the Scala Center: https://www.youtube.com/watch?v=dv7fPmgFTNA and https://www.youtube.com/watch?v=UePrOa_1Am8

There are probably things we can significantly improve for the website. Some open questions:
* How to property "advertize" these tutorials. Currently they are a bit hidden.
* Should we split the Laminar+ST tutorial? It is a bit large in written form.

Comments welcome!

### CODE REVIEW

**Comment by raquo** [comment]
Since the tutorials page is so small, maybe move its contents to the parent https://www.scala-js.org/doc/ ?

As a reference, for Laminar itself, I think our combined learning resources page https://laminar.dev/resources works reasonably well... or at least, nobody has so far spoken up to suggest doing it differently :)

File: doc/tutorial/getting-started/laminar-scalablytyped.md [hunk 1]
```diff
@@ -0,0 +1,808 @@
...
+object DataItem:
+  def apply(): DataItem = DataItem(DataItemID(), "?", Math.random())
+end DataItem
+
+type DataList = List[DataItem]
+{% endhighlight %}
+
+In addition to a `label` and a `value`, we include an `id: DataItemID` in our `DataItem` class.
+The ID will serve to uniquely identify rows of our table even if they happen to have the same label and value.
+Think about a delete button on each row: if we click it, we would like the corresponding row to be removed, not another one with the same content.
```
[the remark this answers is not in the document]
Comment (julienrf) [reply doc/tutorial/getting-started/laminar-scalablytyped.md:213]: (OK I see now that we use the ID below in the `split` call)

**Review by julienrf — COMMENTED** [review]
Wonderful!

I agree that the ScalablyTyped part could be moved out of the Laminar part. But in the end, they talk about the same program so it may require significant work to arrange things to make them more independent of each other… I am fine with keeping them together as well.

File: doc/tutorial/getting-started/index.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
+---
+layout: doc
+title: Getting Started with Scala.js, Vite, Laminar and ScalablyTyped
+---
+
+This series of tutorials teaches you how to use Scala.js together with modern development tools.
```
[the remark this answers is not in the document]
Comment (adpi2) [reply doc/tutorial/getting-started/index.md:6]: Maybe you should move the content of `## Tutorials` up here.

File: doc/tutorial/getting-started/index.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+If you have time, reading and applying them in order will give you more in-depth knowledge about the development environment.
+
+If you are in a hurry, you can skip the ones you are not interested in.
+Each tutorial starts with a link to repo that you can clone to get off the ground.
+
+In any case, make sure that you have the Prerequisites listed below covered.
```
Comment (adpi2) [inline doc/tutorial/getting-started/index.md:13]: I find it harder to read that `# Prerequisites` does not directly follow this sentence

File: doc/tutorial/getting-started/scalajs-vite.md [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+We use Vite to provide live-reloading of the Scala.js application in the browser.
+
+Going through this tutorial will make sure you understand the basic building blocks.
+If you prefer to skip this step and directly write Scala.js code, you may jump to [Getting Started with Scala.js and Laminar](./laminar-scalablytyped.html).
+
+If you prefer to navigate the end result for this tutorial directly, checkout [the scalajs-vite-end-state branch](https://github.com/sjrd/scalajs-sbt-vite-laminar-chartjs-example/tree/scalajs-vite-end-state) instead of creating everything from scratch.
+
+## Prerequisites
+
+Make sure to install [the prerequisites](./index.html) before continuing further.
```
Comment (adpi2) [inline doc/tutorial/getting-started/scalajs-vite.md:16]: Would something like this work?

```suggestion
Make sure to install [the prerequisites](./index.html#prerequisites) before continuing further.
```

File: doc/tutorial/getting-started/scalajs-vite.md [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+
+## Vite template
+
+We bootstrap our setup using the vanilla Vite template.
+Navigate to a directory where you store projects, and run the command
+
+{% highlight shell %}
+$ npm create vite@latest
```
[the remark this answers is not in the document]
Comment (adpi2) [reply doc/tutorial/getting-started/scalajs-vite.md:24]: Indeed, could the latest Vite template break the tutorial?

File: doc/tutorial/getting-started/laminar-scalablytyped.md [hunk 6]
```diff
@@ -0,0 +1,808 @@
...
+title: Getting Started with Laminar and ScalablyTyped
+---
+
+In this second tutorial, we learn how to develop UIs in Scala.js with [Laminar](https://laminar.dev/) and how to integrate JavaScript libraries with [ScalablyTyped](https://scalablytyped.org/).
+
+We start here with the project setup developed in the previous tutorial about [Setting up Scala.js with Vite](./scalajs-vite.html).
+To follow along this tutorial, either use the result of the previous tutorial, or checkout [the scalajs-vite-end-state branch](https://github.com/sjrd/scalajs-sbt-vite-laminar-chartjs-example/tree/scalajs-vite-end-state) of the accompanying repo.
+
+If you prefer to navigate the end result for this tutorial directly, checkout [the laminar-scalablytyped-end-state branch](https://github.com/sjrd/scalajs-sbt-vite-laminar-chartjs-example/tree/laminar-scalablytyped-end-state) instead.
+
```
[the remark this answers is not in the document]
Comment (adpi2) [reply doc/tutorial/getting-started/laminar-scalablytyped.md:12]: Or to integrate the resulting component in this page, like in the [Laminar example pages](https://laminar.dev/examples/hello-world).

File: doc/tutorial/getting-started/laminar-scalablytyped.md [hunk 7]
```diff
@@ -0,0 +1,808 @@
...
+      child.text <-- counter,
+      ...
+    )
+{% endhighlight %}
+
+We now have optimal reuse of `tr` elements based on the `id` of the input `DataItem`s.
+
+[See it in action and fiddle with the code in Scribble](https://scribble.ninja/u/sjrd/cwhugissosbhkucvskpwbqedrntp)
+
+## The actual chart
```
Comment (adpi2) [inline doc/tutorial/getting-started/laminar-scalablytyped.md:375]: This could be a third page of the tutorial, as it solves an independent problem using an independent set of tools.

File: doc/tutorial/getting-started/index.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (sjrd) [reply doc/tutorial/getting-started/index.md:6]: Done as suggested.

File: doc/tutorial/getting-started/laminar-scalablytyped.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (sjrd) [reply doc/tutorial/getting-started/laminar-scalablytyped.md:12]: I will have to investigate how I can do that.

File: doc/tutorial/getting-started/laminar-scalablytyped.md [hunk 8]
```diff
@@ -0,0 +1,808 @@
...
+
+If you prefer to navigate the end result for this tutorial directly, checkout [the laminar-scalablytyped-end-state branch](https://github.com/sjrd/scalajs-sbt-vite-laminar-chartjs-example/tree/laminar-scalablytyped-end-state) instead.
+
+## Prerequisites
+
+Make sure to install [the prerequisites](./index.html) before continuing further.
+
+## Introducing Laminar
+
+To start off, we add a dependency on Laminar in our `build.sbt`:
```
[the remark this answers is not in the document]
Comment (sjrd) [reply doc/tutorial/getting-started/laminar-scalablytyped.md:24]: I added something just above. WDYT?

File: doc/tutorial/getting-started/laminar-scalablytyped.md [hunk 9]
```diff
@@ -0,0 +1,808 @@
...
+      onClick --> { event => counter.update(c => c + 1) },
+    )
+  end counterButton
+{% endhighlight %}
+
+We declare `counter` as a `Var[Int]` initialized with `0`.
+We then use it in two *bindings*:
+
+* In `child.text <-- counter`, we declare a text child of the button whose content will always reflect the value of `counter`.
+  As the value of `counter` changes over time, so does the text in the button.
```
[the remark this answers is not in the document]
Comment (sjrd) [reply doc/tutorial/getting-started/laminar-scalablytyped.md:143]: I added explanations a couple paragraphs below. WDYT?

File: doc/tutorial/index.md [hunk 10]
```diff
@@ -6,6 +6,7 @@ title: Tutorials
 Arguably, the most efficient way to learn new technology is to try it out yourself. Here, we have collected a set
 of tutorials from simple to complete web app, to get you started right away!
 
+* [Getting Started with Scala.js, Vite, Laminar and ScalablyTyped](./getting-started/)
 * [Basic tutorial](./basic/) to get you started
```
[the remark this answers is not in the document]
Comment (sjrd) [reply doc/tutorial/index.md:10]: Yes, this is part of the open questions on how to advertise this tutorial appropriately. The top-level pages probably need to be rethought.

File: doc/tutorial/getting-started/scalajs-vite.md [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+  var counter = 0
+
+  def setCounter(count: Int): Unit =
+    counter = count
+    element.innerHTML = s"count is $counter"
+
+  element.addEventListener("click", e => setCounter(counter + 1))
+  setCounter(0)
+end setupCounter
+{% endhighlight %}
```
[the remark this answers is not in the document]
Comment (sjrd) [reply doc/tutorial/getting-started/scalajs-vite.md]: I added an explanation. WDYT?

File: doc/tutorial/getting-started/laminar-scalablytyped.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (julienrf) [reply doc/tutorial/getting-started/laminar-scalablytyped.md:24]: Great!

File: doc/tutorial/getting-started/scalajs-vite.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (julienrf) [reply doc/tutorial/getting-started/scalajs-vite.md]: Great!

File: doc/tutorial/getting-started/laminar-scalablytyped.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (julienrf) [reply doc/tutorial/getting-started/laminar-scalablytyped.md:143]: Nice, thank you!

**Review by gzm0 — COMMENTED** [review]
Releasing the one comment I had. I won't get to this before a couple of days 🤷

**Comment by oyvindberg** [comment]
Sorry, I'm late to the party here. Thanks for adding this, these are incredibly useful pointers for beginners!

I agree with @julienrf that mixing ST and Laminar in the same tutorial probably should probably be avoided - I think it's fairly unlikely that a beginner interested in one of them is also interested in the other at the same time.

How about we make the Laminar tutorial something more todo-listy, and then extract the `chart.js` / ST pieces into its own tutorial? I can contribute some hours towards extracting the ST parts into it's own page.

**Comment by sjrd** [comment]
I have finally found enough time to decompose the tutorial. There is now a Laminar-only tutorial, and then a separate one for ScalablyTyped. The Laminar tutorial is entirely available on Scribble for people to see the end result and play with it.

File: doc/tutorial/getting-started/index.md [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+title: Getting Started Tutorial Series
+---
+
+This series of tutorials teaches you how to use Scala.js together with modern development tools.
+
+1. [Scala.js and Vite](./scalajs-vite.html):
+  * Set up a hello world project ready for live reloading in the browser.
+  * Generate minimized production assets.
+2. [Laminar](./laminar.html):
+  * Build UIs with Laminar using Function Reactive Programming (FRP), a hybrid model between imperative and functional programming particularly well suited for UI development in Scala.
```
Comment (Sporarum) [inline doc/tutorial/getting-started/index.md:12]:
```suggestion
  * Build UIs with Laminar using Functional Reactive Programming (FRP), a hybrid model between imperative and functional programming particularly well suited for UI development in Scala.
```

File: doc/tutorial/getting-started/index.md [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+2. [ScalablyTyped](./scalablytyped.html):
```
Comment (Sporarum) [inline doc/tutorial/getting-started/index.md:13]:
```suggestion
3. [ScalablyTyped](./scalablytyped.html):
```
(Probably doesn't change anything to the generated document)

File: doc/tutorial/getting-started/index.md [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+  * Integrate JavaScript libraries using ScalablyTyped.
+
+If you have time, reading and applying them in order will give you more in-depth knowledge about the development environment.
+
+If you are in a hurry, you can skip the ones you are not interested in.
+Each tutorial starts with a link to a repo that you can clone to get off the ground.
+
+In any case, make sure that you have the Prerequisites listed below covered.
```
Comment (Sporarum) [inline doc/tutorial/getting-started/index.md:21]:
```suggestion
In any case, make sure that you have the prerequisites listed below covered.
```

File: doc/tutorial/getting-started/scalablytyped.md [hunk 15]
```diff
@@ -0,0 +1,274 @@
...
+This prevents memory leaks.
+
+Our application now properly renders the data model as a chart.
+When we add or remove data items, the chart is automatically updated, thanks to the connection established by the `dataSignal -->` binder.
+
+## Conclusion
+
+That concludes our tutorial on ScalablyTyped.
+
+Finally, we saw how to configure ScalablyTyped to get static types for external JavaScript libraries, and how to integrate a third-party component into a Laminar model.
```
Comment (Sporarum) [inline doc/tutorial/getting-started/scalablytyped.md:274]:
```suggestion
We saw how to configure ScalablyTyped to get static types for external JavaScript libraries, and how to integrate a third-party component into a Laminar model.
```
The "Finally" feels weird since there is a sentence before, I interpreted it as "At last, and not as "To conclude" (as I presume was intended)

**Review by Sporarum — COMMENTED** [review]
I have minor stylistic comments, but apart from that, it looks good to me !

File: doc/tutorial/laminar.md [hunk 16]
```diff
@@ -0,0 +1,675 @@
...
+
+Let us isolate the button definition in a separate method, and illustrate how to write its behavior using Laminar:
+
+{% highlight scala %}
+  def counterButton(): Element =
+    val counter = Var(0)
+    button(
+      tpe := "button",
+      "count is ",
+      child.text <-- counter,
```
Comment (gzm0) [inline doc/tutorial/laminar.md:135]: I find this confusing: Is the append implicit / does it just add two text nodes?

File: doc/tutorial/laminar.md [hunk 17]
```diff
@@ -0,0 +1,675 @@
...
+        valueStr.toDoubleOption.foreach(valueUpdater.onNext)
+      },
+    )
+  end inputForDouble
+{% endhighlight %}
+
+It is more complicated than the one for `String`s because of the need for parsing and formatting.
+This complexity perhaps better highlights the benefit of encapsulating it in a dedicated method (or component).
+
+We leave it to the reader to understand the details of this method.
```
Comment (gzm0) [inline doc/tutorial/laminar.md:597]: The one thing I do not understand even after reading carefully: Why are the bindings between valueSignal / strValue / valueUpdater part of the `input` element?

File: doc/tutorial/scalablytyped.md [hunk 18]
```diff
@@ -0,0 +1,274 @@
...
+
+We set up our new dependencies as follows.
+
+In `project/plugins.sbt`, we add a dependency on ScalablyTyped:
+
+{% highlight scala %}
+addSbtPlugin("org.scalablytyped.converter" % "sbt-converter" % "1.0.0-beta40")
+{% endhighlight %}
+
+In `package.json`, we add dependencies on Chart.js, its TypeScript type definitions, and on the TypeScript compiler, which is required by ScalablyTyped:
```
Comment (gzm0) [inline doc/tutorial/scalablytyped.md:34]: Consider providing the relevant `npm` invocations rather than suggesting to change the file. IIRC, doing this does not deal well with lock files.

File: doc/tutorial/scalablytyped.md [hunk 19]
```diff
@@ -0,0 +1,274 @@
...
+This `-->` binder allows to bridge the FRP world of `dataSignal` with the imperative world of Chart.js.
+
+Note that we put the `-->` binder as an argument to the Laminar `canvas` element.
+This may seem suspicious, as neither `dataSignal` nor the callback have any direct relationship to the DOM canvas element.
+We do this to tie the lifetime of the binder to the lifetime of the `canvas` element.
+When the latter gets unmounted, we release the binder connection, possibly allowing resources to be reclaimed.
+
+In general, every binder must be *owned* by a Laminar element.
+It only gets *activated* when that element is mounted.
+This prevents memory leaks.
```
Comment (gzm0) [inline doc/tutorial/scalablytyped.md:265]: Ah, these two paragraphs are the answer to my original question. Maybe move it? Or cross reference?

**Review by gzm0 — CHANGES_REQUESTED** [review]
All-in-all this looks very nice. I have left minor comments. However, two bigger things stand out:

- There is no testing. IMO the vite tutorial and the laminar tutorial both need some input there (vite: how to set it up, laminar: how to test RFP).
- The build integration with vite seems suboptimal. OK to try it out, not good enough to recommend it as part of a tutorial IMO.

File: doc/tutorial/scalajs-vite.md [hunk 20]
```diff
@@ -0,0 +1,416 @@
...
+  if (result.error)
+    throw result.error;
+  if (result.status !== 0)
+    throw new Error(`sbt process failed with exit code ${result.status}`);
+  return result.stdout.toString('utf8').trim();
+}
+
+// Get the output of fastLinkJS or fullLinkJS depending on isDev()
+const scalaJSOutputTask = isDev() ? "fastLinkJSOutput" : "fullLinkJSOutput";
+const scalaJSOutput = printSbtTask(scalaJSOutputTask);
```
[the remark this answers is not in the document]
Comment (sjrd) [reply doc/tutorial/scalajs-vite.md:330]: > Have you considered a vite / rollup plugin (potentially in-place)? I realize that this is potentially significant scope creep, but I think for the first tutorial on the main page we should be better than that.

I have considered that. It was one of my first ideas. But then I internally got as critique that it would not help Mill users (or other build tools). As is, the Mill user can probably figure out that they can invoke Mill instead of sbt. If we make a vite-sbt plugin, this is hidden and not available to Mill users.

That said, you're not the first one also saying that this looks jarring as getting-started experience. So I'm taking it as a sign that it's probably worth doing the vite-sbt integration better, at the cost of being less friendly to other build tools.

File: doc/tutorial/laminar.md (hunk 16 above)
Comment (sjrd) [reply doc/tutorial/laminar.md:135]: It creates two text nodes.

I can also use `child.text <-- counter.map(c => s"count is $c)`, but that requires to already explain `map` at that point, which doesn't fit the flow of the tutorial.

File: doc/tutorial/scalajs-vite.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (gzm0) [reply doc/tutorial/scalajs-vite.md:330]: > That said, you're not the first one also saying that this looks jarring as getting-started experience. So I'm taking it as a sign that it's probably worth doing the vite-sbt integration better, at the cost of being less friendly to other build tools.

FWIW: If the plugin is reasonably simple, it might make sense to go over how it works in the tutorial. So it's easier to port it to mill.

Or even inline the plugin into the build (if it can go into a separate file, it's maybe not too bad and it'll make it much easier to, say, deal with multiple Scala.js outputs for say FE / BE).

**Comment by sjrd** [comment]
OK, I have finally updated everything to use the published [`vite-plugin-scalajs`](https://github.com/scala-js/vite-plugin-scalajs). I have also addressed other comments, such as adding a testing section in the Laminar tutorial (I did not do it in the Vite setup to keep it to the minimum).

This is ready for a new round of review.

File: doc/tutorial/scalajs-vite.md [hunk 21]
```diff
@@ -0,0 +1,416 @@
...
+
+import org.scalajs.dom
+
+// import javascriptLogo from "/javascript.svg"
+@js.native @JSImport("/javascript.svg", JSImport.Default)
+val javascriptLogo: String = js.native
+
+@main
+def LiveChart(): Unit =
+  dom.document.querySelector("#app").innerHTML = s"""
```
[the remark this answers is not in the document]
Comment (sjrd) [reply doc/tutorial/scalajs-vite.md:167]: Yes, that's unfortunate. But I think we have to live with it for now. Hopefully this gets addressed upstream and we can upgrade at some point.

File: doc/tutorial/scalajs-vite.md
Comment (julienrf) [inline doc/tutorial/scalajs-vite.md:1]: This looks great to me! There is maybe one thing that could be simplified though. We start with a Vite template where we “undo” some things (e.g. we remove the .js files) before we set up Scala.js. I wonder if it would be simpler to provide a template that has the Scala.js integration from the beginning?

**Review by julienrf — APPROVED** [review]
This looks good, but I wonder if we can simplify even further the first tutorial. See my comment below.

File: doc/tutorial/scalajs-vite.md [hunk 22]
```diff
@@ -0,0 +1,357 @@
...
+export default defineConfig({
+  plugins: [scalaJSPlugin()],
+});
+{% endhighlight %}
+
+Finally, open the file `main.js`, remove almost everything to leave only the following two lines:
+
+{% highlight javascript %}
+import './style.css'
+import 'scalajs:main.js'
```
[the remark this answers is not in the document]
Comment (sjrd) [reply doc/tutorial/scalajs-vite.md:270]: Done.

File: doc/tutorial/scalajs-vite.md
Comment (sjrd) [reply doc/tutorial/scalajs-vite.md:1]: It's a possibility, of course. That would significantly change the flow of the tutorial. The idea for that was that the example repo branches would address this:

> If you prefer to navigate the end result for this tutorial directly, checkout [the scalajs-vite-end-state branch](https://github.com/sjrd/scalajs-sbt-vite-laminar-chartjs-example/tree/scalajs-vite-end-state) instead of creating everything from scratch.

The next tutorial about Laminar also suggests to directly clone that branch to start off.

**Review by gzm0 — APPROVED** [review]
LGTM. Only minor comments from my side.

File: doc/tutorial/laminar.md [hunk 23]
```diff
@@ -0,0 +1,782 @@
...
+{% highlight diff %}
+     // Depend on Laminar
+     libraryDependencies += "com.raquo" %%% "laminar" % "0.14.2",
++
++    // Testing framework
++    libraryDependencies += "org.scalameta" %%% "munit" % "0.7.29" % Test,
+   )
+{% endhighlight %}
+
+After re-importing the project in the IDE (which should be prompted), we create a new file `src/test/scala/livechart/ModelTest.scala`.
```
[the remark this answers is not in the document]
Comment (sjrd) [reply doc/tutorial/laminar.md]: I removed the `(...)`. It was not really necessary. If the IDE prompts, the user will see it; and if it doesn't for some reason, they won't puzzle over the fact that it doesn't.

**Comment by sjrd** [comment]
Thanks for the review :D Let's ship it 🚀
